### ISSUE TITLE
Avoid recompilation caused by variably shaped arrays?

### ISSUE THREAD

**Issue by john-heyer** [google/jax#2521]
I'm working with a function that computes a likelihood given a 1D array. The problem is that this array has variable length, and so the `jit`'ed function triggers recompilation when the length of this array changes, which happens frequently in my program, causing a significant hit in performance.

I know that there are various levels of abstraction, and see the existence of an `Unshaped` arrays in the docs: "JAX can also trace at higher levels of abstraction, like Unshaped, but that’s not currently the default for any transformation".

First, is it even possible with JAX to compile without declaring the shape a compile-time constant? Would using `Unshaped` arrays solve this? If so, how can I tell JAX to trace with this level of abstraction?

Thank you!

**Comment by mattjj** [comment]
Another great question!

There are two separate issues here worth disentangling:
1. staging programs out of Python by trace-specializing, and
2. compiling those staged-out programs using XLA.

The levels of abstraction in JAX pertain to the former. It's true that JAX can trace-specialize at the Unshaped level, but semi-tangentially there's an interesting tradeoff between amount of abstraction and the restrictions required for Python code to be traceable. For example, we couldn't trace `mean0 = lambda x: x.sum(0) / x.shape[0]` at the Unshaped level, while if we specialize on shape that Python function is traceable. (We could trace `lambda x: np.sin(np.cos(x))` at Unshaped just fine though.) There's a relationship to reverse-mode autodiff too: we can transpose `lambda x: x.sum(0)` specialized on shape, but not unshaped, because we need to know what size to broadcast to. (We could transpose `lambda x: x * 5.` though.) Just from the point of view of this tradeoff, it's been a nice sweet spot to specialize on shapes.

But you asked about compilation in particular. Let's set issue 1 aside (say we have a Python function that we can specialize and transform at the Unshaped level, like `lambda x: np.sin(np.cos(x))`). Once we have a staged-out program (i.e. a jaxpr) that is only specialized at the Unshaped level, could we compile one XLA program for it, and thus not have to recompile for different shapes?

The answer is no: XLA HLO programs include concrete shapes in their type system. That is, just for issue 2 and unrelated to issue 1, we'd need to specialize the staged-out Unshaped program on argument shapes to be able to generate an XLA HLO program for it, and we'd have to re-invoke the compiler for each new shape specialization. XLA has its own reasons for specializing on shape: for instance, shape-specialized programs can be statically memory allocated and layout-optimized, and decisions like whether to fuse certain operations or rematerialize intermediates depend entirely on the shapes/sizes involved. That's what gives XLA its incredible optimization power, especially on an accelerator like a TPU.

You could say you're willing to give up on some of those shape-specialized optimization opportunities if it meant you didn't have to recompile as often. That makes sense! (In fact, on CPU and GPU we're used to not compile-time specializing on shapes for GEMM kernels, though they may be selected at runtime based on argument shapes.) People are exploring array-oriented compilers at those different points in the design space, and it's exciting stuff. I expect JAX will take advantage of these kinds of capabilities as they emerge, in XLA and/or related technologies in the MLIR world. But for now the best compiler (XLA) requires shape-specialized programs. (You could actually lower an Unshaped-specialized jaxpr to a TensorFlow program pretty effectively, though I'm not sure if the TF graph executor (aka interpreter) has the right performance characteristics you're looking for. Does it?)

But wait! We have a secret weapon here! JAX, transform!

One of the most exciting new JAX transformations is `mask`. It's a prototype that needs a few weeks more work, and of course it's totally undocumented as per our custom, but it can help us because `mask` can let us automatically implement shape-polymorphic semantics in shape-monomorphic programs. Here's what I mean:

```python
import jax.numpy as np
from jax import jit, mask

def bucket_jit(f):
  compiled_f = jit(mask(f, ['n'], ''))
  def wrapped(x):
    amount = 128 - x.shape[0] % 128
    padded_x = np.pad(x, (0, amount))
    return compiled_f([padded_x], dict(n=x.shape[0]))
  return wrapped

@bucket_jit
def foo(x):
  print("recompiling!")  # actually retracing, but effectively correct
  return np.tanh(np.sum(x))

foo(np.arange(4))  # recompiling!
foo(np.arange(5))
foo(np.arange(6))
foo(np.arange(300))  # recompiling!

from jax import grad
grad(foo)(np.arange(3.))  # recompiling!
grad(foo)(np.arange(4.))
grad(foo)(np.arange(129.))  # recompiling!
```

I'm not sure I'd recommend you use `mask` quite yet, but this is one example of what it can do. (There's more! While we can do `jit` + `mask` to avoid recompilations, we can do `vmap`+`mask` to do ragged batching, e.g. for RNNs. We can also use the same machinery to do import-time shape checking of your code.) There are some examples in [the tests](https://github.com/google/jax/blob/master/tests/masking_test.py).

We realized at one point that Unshaped (which doesn't specialize on any shape information or even rank) was way too big a leap away from Shaped to be useful. Instead, `mask` specializes on partial shape information, in particular on parametrically polymorphic dimensions (which can be concrete too: do as much specialization as you like).

The `mask` transform doesn't do anything you couldn't do by hand. That is, you can pad and mask things by hand so as to implement shape-polymorphic semantics in a shape-monomorphic way. But it's a big pain which scales badly with the complexity of the programs you want to handle. It's just like autodiff: sure you can write derivative code by hand, but it's much better to have an automatic, composable transformation do it for you.

Anyway, here's the short version of answers to your questions:

> First, is it even possible with JAX to compile without declaring the shape a compile-time constant?

No, XLA HLO programs we compile must be shape-specialized, even though JAX can stage some programs out of Python based on less specialization.

> Would using Unshaped arrays solve this?

No, because while it would avoid trace-specializing on shapes, we have no way to compile a single program with no shape specialization. (Forward-looking, it's probably better to specialize on _some_ shape information, even if just parametric polymorphism.)

> If so, how can I tell JAX to trace with this level of abstraction?

There's no user API for that, though if you want to fiddle with it for fun you can [change `make_jaxpr` not to use `xla.abstractify`](https://github.com/google/jax/blob/42dbfd43d4b219dd31a102d63b6fb241acbb2426/jax/api.py#L1399) (which is effectively just `lambda x: raise_to_shaped(core.get_aval(x))`) and instead abstract to the Unshaped level (maybe like `lambda x: Unshaped(core.get_aval(x).dtype)`).

Hope that answers your question, though unfortunately I'm answering it in the negative. You could play with `mask`, but I suspect it might be tricky until we post at least minimal documentation, and even then we don't know whether it'll work well for your use case (are you on CPU or GPU?).

I'm going to close this issue on the hunch that this was enough of an answer, but please reopen (or open new ones) as needed. Questions are always welcome!

**Comment by mattjj** [comment]
One more thing: if your function is reasonably self-contained, post it here and maybe we can use it as a motivating example for `mask`!

**Comment by john-heyer** [comment]
Hi Matt, thanks so much again for the incredible response (I also noticed you were an MIT grad student, I'm an MENG currently)!

This information cleared up a lot. `mask` seems like it'd be very useful for me (I am on CPU with lots of cores), so I'm excited for the documentation!

My function is not self-contained at all, but the work I am doing uses [Reversible Jump](https://en.wikipedia.org/wiki/Reversible-jump_Markov_chain_Monte_Carlo) MCMC, which allows inference when the number of model parameters is unknown (hence the constantly varying size of the array passed to the likelihood function!).

Once the chain has warmed up, there is a typical set where `n` (the number of parameters and the size of the array) hangs out. So, I was thinking of doing something hacky like having a pre-sampling step where I essentially `jit` many versions of the same function with various shapes. Of course this doesn't scale well, but it might help me out for my experiments. Does this sound reasonable? Does JAX already cache these different "versions" of the `jit`ed functions? Do you have a better idea?

Thank you again Matt!

(I reopened the issue as you suggested, but feel free to close as your answer was what I was looking for).

**Comment by mattjj** [comment]
> Does this sound reasonable? Does JAX already cache these different "versions" of the jited functions? Do you have a better idea?

Yes, that'll work; JAX indeed caches compiled versions of functions. In fact, I'd say this is the canonical way to do it. (Fancier tracing JIT compilers, like from JavaScript, have more fancy trace formation and management, but in JAX you're basically doing it manually and explicitly. The upside is fewer performance surprises and a simpler implementation.)

The way `mask`+`jit` works, with XLA seeing a shape-monomorphic program that happens to have padding and masking logic built into it, is great on SIMD-style accelerators like TPU and GPU, but not very smart on CPU. It'd be better if we could lower our shape polymorphic information to the compiler, and have it do something smart in a backend-dependent way. Unfortunately we don't have the right compiler tools for the job yet, though we're starting to advocate for compiler improvements like this. (XLA:CPU is actually the least developed backend, and in particular it has longer compile times and is less optimized than the GPU and especially TPU backends. It's still the best game in town that I know of, though.)

If you can afford doing all the compilation then that's the way to go for now. If you can't, you can use `mask` and shape bucketing to do less compilation, but that'll likely sacrifice some runtime performance (because the masking computations cost something). Other than that, we can all advocate for more array-oriented CPU compiler development!

> (I also noticed you were an MIT grad student, I'm an MENG currently)!

That's right! It's great to have these little connections. I first met many of my present-day collaborators while I was there.

**Comment by john-heyer** [comment]
Makes sense! Thanks for all the help, I'm sure I'll be back!

**Comment by john-heyer** [comment]
One last question relevant to this thread: how "smart" is the JAX tracing?

What I mean is, say I have a highest-level `sample` method, that I can easily `jit`, whose arguments never change and so will never recompile. However, at each iteration, the sampler calls a `likelihood` function whose argument changes shape (described above), and will retrace often. Will this trigger a retracing of the entire `sample` method, or does JAX know to only recompile necessary parts?

The opposite example is also found in my code: the `likelihood` function essentially sums up the individual likelihoods `likelihood_1 ... likelihood_n`, and so the subfunction, i'll call `single_likelihood`, should never need to be recompiled, while the outer function will upon changing shape. Would JAX recompile everything here? (It's still worth compiling the outer function as we're summing up many matrices).

The performance impact may be insignificant, I'm not really sure! I guess what I'm asking is if I should write my code in a "smart" way to avoid unnecessary recompilation, or if it's as easy as calling `jit` at the highest level, and letting JAX handle the rest.

**Comment by mattjj** [comment]
Hm, to check in detail it might be worth setting up a toy example (in a new issue, so it doesn't get missed). But in general JAX's `jit` is all-or-nothing: either all the computation under the `jit`-decorated function gets staged out and compiled into one XLA computation (for each set of input argument shapes), or it'll fail loudly. Sub-components don't get compiled separately.

So if you can `jit` a function, you don't have to worry about any recompilation underneath it.

**Comment by gnool** [comment]
@mattjj just curious, is the `mask` transformation documented somewhere that I can refer to? I am similarly facing the problem that one of my function arguments is dynamic in array shape, thus triggering recompilation. Hopefully `mask` could help!

**Comment by mattjj** [comment]
It's not documented yet, but it will be when it's ready for any general use. I think it'll happen in 2020, but I have been wrong before...

**Comment by gnool** [comment]
@mattjj A follow-up question to the toy example given above. Suppose I have a function f that returns an array of dynamic size determined by one of the function argument `indices` which is a 1D int array. Effectively, the function would return foo[indices], where `foo` itself is an array to be sliced. Can I use mask for this purpose to avoid recompilation?

**Comment by mattjj** [comment]
Yes that's part of the design, but no not in its current implementation. Of course, it's all vaporware and you shouldn't believe anything I say until you can actually use it :P

Here's a slide from a talk I gave on it last week. The middle two examples, under (2), are versions of exactly this use case.

![image](https://user-images.githubusercontent.com/1458824/96543027-55276800-1258-11eb-9b44-2a7780ee67e1.png)

**Comment by gnool** [comment]
Looks like this is happening very actively behind the scene, will definitely keep an eye out for upcoming updates.

For now I'll make the function return a constant size array (and knowing that only some of the array elements are getting updated by `indices`).

**Comment by proteneer** [comment]
Bump - any updates on this?
